flyway baseline-version has no effect

### Describe the bug

After updating the config key quarkus.flyway.baseline-version: "2022.09.02.2", I was expecting newly created environments to use 2022.09.02.2 as first version (with migrate-at-start enabled). 

It turns out that migrate-at-start does not execute the baseline operation and all migrations before 2022.09.02.2 are still executed.

### Expected behavior

first migration of newly created environments is equal to baseline-version

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.12.0.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

@gastaldi 
It seems that org.flywaydb.core.Flyway bean cannot be accessed inside the callback because it is called during static init 
(see the stacktrace bellow, I tried to access it with CDI.current().select(Flyway.class).get()).

So I will stick to the startup event observer for now but it is not really a proper solution.

```
Caused by: org.flywaydb.core.api.FlywayException: Error while executing beforeMigrate callback: Synthetic bean instance for org.flywaydb.core.Flyway not initialized yet: org_flywaydb_core_Flyway_3cb728d15a0a04006cc75d03f784feb2bba9a5fb
	- a synthetic bean initialized during RUNTIME_INIT must not be accessed during STATIC_INIT
	- RUNTIME_INIT build steps that require access to synthetic beans initialized during RUNTIME_INIT should consume the SyntheticBeansRuntimeInitBuildItem
	at org.flywaydb.core.internal.callback.DefaultCallbackExecutor.handleEvent(DefaultCallbackExecutor.java:136)
	at org.flywaydb.core.internal.callback.DefaultCallbackExecutor.execute(DefaultCallbackExecutor.java:129)
	at org.flywaydb.core.internal.callback.DefaultCallbackExecutor.lambda$execute$0(DefaultCallbackExecutor.java:116)
	at org.flywaydb.core.internal.jdbc.TransactionalExecutionTemplate.execute(TransactionalExecutionTemplate.java:55)
	at org.flywaydb.core.internal.callback.DefaultCallbackExecutor.execute(DefaultCallbackExecutor.java:114)
	at org.flywaydb.core.internal.callback.DefaultCallbackExecutor.onMigrateOrUndoEvent(DefaultCallbackExecutor.java:70)
	at org.flywaydb.core.internal.command.DbMigrate.migrate(DbMigrate.java:84)
	at org.flywaydb.core.Flyway$1.execute(Flyway.java:176)
	at org.flywaydb.core.Flyway$1.execute(Flyway.java:126)
	at org.flywaydb.core.FlywayExecutor.execute(FlywayExecutor.java:204)
	at org.flywaydb.core.Flyway.migrate(Flyway.java:126)
	at io.quarkus.flyway.runtime.FlywayRecorder.doStartActions(FlywayRecorder.java:76)
	at io.quarkus.deployment.steps.FlywayProcessor$createBeansAndStartActions2063183959.deploy_0(Unknown Source)
	at io.quarkus.deployment.steps.FlywayProcessor$createBeansAndStartActions2063183959.deploy(Unknown Source)
	... 68 more
```

--- a/extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayExtensionBaselineAtStartExistingSchemaHistoryTableTest.java
+++ b/extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayExtensionBaselineAtStartExistingSchemaHistoryTableTest.java
@@ -1,0 +1,40 @@
+package io.quarkus.flyway.test;
+
+import static org.junit.jupiter.api.Assertions.assertNull;
+
+import jakarta.inject.Inject;
+
+import org.flywaydb.core.Flyway;
+import org.flywaydb.core.api.MigrationInfo;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+public class FlywayExtensionBaselineAtStartExistingSchemaHistoryTableTest {
+    @Inject
+    Flyway flyway;
+
+    static final FlywayH2TestCustomizer customizer = FlywayH2TestCustomizer
+            .withDbName("quarkus-baseline-at-start-existing-schema-history")
+            .withPort(11309)
+            .withInitSqlFile("src/test/resources/h2-init-schema-history-table.sql");
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .setBeforeAllCustomizer(customizer::startH2)
+            .setAfterAllCustomizer(customizer::stopH2)
+            .withApplicationRoot((jar) -> jar
+                    .addClass(FlywayH2TestCustomizer.class)
+                    .addAsResource("db/migration/V1.0.0__Quarkus.sql")
+                    .addAsResource("baseline-at-start-existing-schema-history-table-config.properties",
+                            "application.properties"));
+
+    @Test
+    @DisplayName("Baseline at start is not executed against existing schema-history-table")
+    public void testFlywayConfigInjection() {
+        MigrationInfo migrationInfo = flyway.info().current();
+        assertNull(migrationInfo, "Flyway baseline was executed on existing schema history table");
+    }
+}

--- a/extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayExtensionBaselineAtStartTest.java
+++ b/extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayExtensionBaselineAtStartTest.java
@@ -1,0 +1,39 @@
+package io.quarkus.flyway.test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import jakarta.inject.Inject;
+
+import org.flywaydb.core.Flyway;
+import org.flywaydb.core.api.MigrationInfo;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+public class FlywayExtensionBaselineAtStartTest {
+    @Inject
+    Flyway flyway;
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .withApplicationRoot((jar) -> jar
+                    .addAsResource("db/migration/V1.0.0__Quarkus.sql")
+                    .addAsResource("baseline-at-start-config.properties", "application.properties"));
+
+    @Test
+    @DisplayName("Baseline at start is executed against empty schema")
+    public void testFlywayConfigInjection() {
+        MigrationInfo migrationInfo = flyway.info().current();
+        assertNotNull(migrationInfo, "No Flyway migration was executed");
+        assertTrue(migrationInfo.getType().isBaseline(), "Flyway migration is not a baseline");
+        String currentVersion = migrationInfo
+                .getVersion()
+                .toString();
+
+        assertEquals("1.0.1", currentVersion);
+    }
+}

--- a/extensions/flyway/deployment/src/test/resources/baseline-at-start-config.properties
+++ b/extensions/flyway/deployment/src/test/resources/baseline-at-start-config.properties
@@ -1,0 +1,8 @@
+quarkus.datasource.db-kind=h2
+quarkus.datasource.username=sa
+quarkus.datasource.password=sa
+quarkus.datasource.jdbc.url=jdbc:h2:tcp://localhost/mem:quarkus-baseline-at-start;DB_CLOSE_DELAY=-1
+
+# Flyway config properties
+quarkus.flyway.baseline-at-start=true
+quarkus.flyway.baseline-version=1.0.1

--- a/extensions/flyway/deployment/src/test/resources/baseline-at-start-existing-schema-history-table-config.properties
+++ b/extensions/flyway/deployment/src/test/resources/baseline-at-start-existing-schema-history-table-config.properties
@@ -1,0 +1,8 @@
+quarkus.datasource.db-kind=h2
+quarkus.datasource.username=sa
+quarkus.datasource.password=sa
+quarkus.datasource.jdbc.url=jdbc:h2:tcp://localhost:11309/mem:quarkus-baseline-at-start-existing-schema-history;DB_CLOSE_DELAY=-1
+
+# Flyway config properties
+quarkus.flyway.baseline-at-start=true
+quarkus.flyway.baseline-version=1.0.1

--- a/extensions/flyway/deployment/src/test/resources/h2-init-schema-history-table.sql
+++ b/extensions/flyway/deployment/src/test/resources/h2-init-schema-history-table.sql
@@ -1,0 +1,13 @@
+CREATE TABLE "flyway_schema_history" (
+    "installed_rank" integer NOT NULL,
+    "version" character varying(50),
+    "description" character varying(200) NOT NULL,
+    "type" character varying(20) NOT NULL,
+    "script" character varying(1000) NOT NULL,
+    "checksum" integer,
+    "installed_by" character varying(100) NOT NULL,
+    "installed_on" timestamp without time zone DEFAULT now() NOT NULL,
+    "execution_time" integer NOT NULL,
+    "success" boolean NOT NULL,
+    CONSTRAINT flyway_schema_history_pk PRIMARY KEY ("installed_rank")
+);

--- a/extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayContainer.java
+++ b/extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayContainer.java
@@ -5,6 +5,8 @@ import org.flywaydb.core.Flyway;
 public class FlywayContainer {
 
     private final Flyway flyway;
+
+    private final boolean baselineAtStart;
     private final boolean cleanAtStart;
     private final boolean migrateAtStart;
     private final boolean repairAtStart;
@@ -15,10 +17,11 @@ public class FlywayContainer {
     private final boolean createPossible;
     private final String id;
 
-    public FlywayContainer(Flyway flyway, boolean cleanAtStart, boolean migrateAtStart, boolean repairAtStart,
-            boolean validateAtStart,
+    public FlywayContainer(Flyway flyway, boolean baselineAtStart, boolean cleanAtStart, boolean migrateAtStart,
+            boolean repairAtStart, boolean validateAtStart,
             String dataSourceName, boolean hasMigrations, boolean createPossible) {
         this.flyway = flyway;
+        this.baselineAtStart = baselineAtStart;
         this.cleanAtStart = cleanAtStart;
         this.migrateAtStart = migrateAtStart;
         this.repairAtStart = repairAtStart;
@@ -31,6 +34,10 @@ public class FlywayContainer {
 
     public Flyway getFlyway() {
         return flyway;
+    }
+
+    public boolean isBaselineAtStart() {
+        return baselineAtStart;
     }
 
     public boolean isCleanAtStart() {

--- a/extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayContainerProducer.java
+++ b/extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayContainerProducer.java
@@ -51,7 +51,8 @@ public class FlywayContainerProducer {
         final Flyway flyway = new FlywayCreator(matchingRuntimeConfig, matchingBuildTimeConfig, matchingConfigCustomizers(
                 configCustomizerInstances, dataSourceName)).withCallbacks(callbacks)
                 .createFlyway(dataSource);
-        return new FlywayContainer(flyway, matchingRuntimeConfig.cleanAtStart, matchingRuntimeConfig.migrateAtStart,
+        return new FlywayContainer(flyway, matchingRuntimeConfig.baselineAtStart, matchingRuntimeConfig.cleanAtStart,
+                matchingRuntimeConfig.migrateAtStart,
                 matchingRuntimeConfig.repairAtStart, matchingRuntimeConfig.validateAtStart,
                 dataSourceName, hasMigrations,
                 createPossible);

--- a/extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java
+++ b/extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java
@@ -152,10 +152,21 @@ public final class FlywayDataSourceRuntimeConfig {
     public boolean validateAtStart;
 
     /**
-     * Enable the creation of the history table if it does not exist already.
+     * true to execute Flyway baseline before migrations This flag is ignored if the flyway_schema_history table exists in the
+     * current schema or if the current schema is empty.
+     * Note that this will not automatically call migrate, you must either enable baselineAtStart or programmatically call
+     * flyway.migrate().
      */
     @ConfigItem
     public boolean baselineOnMigrate;
+
+    /**
+     * true to execute Flyway baseline automatically when the application starts.
+     * This flag is ignored if the flyway_schema_history table exists in the current schema.
+     * This will work even if the current schema is empty.
+     */
+    @ConfigItem
+    public boolean baselineAtStart;
 
     /**
      * The initial baseline version.

--- a/extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRecorder.java
+++ b/extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRecorder.java
@@ -11,8 +11,16 @@ import jakarta.enterprise.inject.Default;
 import jakarta.enterprise.inject.UnsatisfiedResolutionException;
 
 import org.flywaydb.core.Flyway;
+import org.flywaydb.core.FlywayExecutor;
 import org.flywaydb.core.api.callback.Callback;
 import org.flywaydb.core.api.migration.JavaMigration;
+import org.flywaydb.core.api.output.BaselineResult;
+import org.flywaydb.core.internal.callback.CallbackExecutor;
+import org.flywaydb.core.internal.database.base.Database;
+import org.flywaydb.core.internal.database.base.Schema;
+import org.flywaydb.core.internal.jdbc.StatementInterceptor;
+import org.flywaydb.core.internal.resolver.CompositeMigrationResolver;
+import org.flywaydb.core.internal.schemahistory.SchemaHistory;
 import org.jboss.logging.Logger;
 
 import io.quarkus.agroal.runtime.DataSources;
@@ -101,12 +109,33 @@ public class FlywayRecorder {
             if (flywayContainer.isValidateAtStart()) {
                 flywayContainer.getFlyway().validate();
             }
+            if (flywayContainer.isBaselineAtStart()) {
+                new FlywayExecutor(flywayContainer.getFlyway().getConfiguration())
+                        .execute(new BaselineCommand(flywayContainer.getFlyway()), true, null);
+            }
             if (flywayContainer.isRepairAtStart()) {
                 flywayContainer.getFlyway().repair();
             }
             if (flywayContainer.isMigrateAtStart()) {
                 flywayContainer.getFlyway().migrate();
             }
+        }
+    }
+
+    static class BaselineCommand implements FlywayExecutor.Command<BaselineResult> {
+        BaselineCommand(Flyway flyway) {
+            this.flyway = flyway;
+        }
+
+        final Flyway flyway;
+
+        @Override
+        public BaselineResult execute(CompositeMigrationResolver cmr, SchemaHistory schemaHistory, Database d,
+                Schema defaultSchema, Schema[] s, CallbackExecutor ce, StatementInterceptor si) {
+            if (!schemaHistory.exists()) {
+                return flyway.baseline();
+            }
+            return null;
         }
     }
 }

